File: 005/Solution.java
/**
 * Solution
 */
public class Solution {

    public String longestPalindrome(String s) {
        String res = "";
        if (s == null || s.length() == 0)
            return res;
        int maxLen = 0;
        boolean[][] palindromeQ = new boolean[s.length()][s.length()];
        for (int i = s.length() - 1; i >= 0; i--) {
            for (int j = i + 1; j < s.length(); j++) {
                if (s.charAt(i) == s.charAt(j) && (j-i <= 2 || palindromeQ[i+1][j-1])) {
                    palindromeQ[i][j] = true;
                    if (j - i + 1 > maxLen) {
                        maxLen = j - i + 1;
                        res = s.substring(i, j+1);
                    }
                }
            }
        }
        return res;
    }

    public static void main(String[] args) {
        String s = "babad";
        Solution sol = new Solution();
        System.out.println(sol.longestPalindrome(s));
    }
}

File: 005/Solution.scala
object Solution {
    def longestPalindrome(s: String): String = {
        if (s.length < 2) {
            return s
        }

        var (i, start, maxLen) = (0, 0, 1)
        while (i < s.length && s.length - i > maxLen / 2) {
            var (left, right) = (i, i)
            while (right < s.length - 1 && s(right+1) == s(right)) {
                right += 1
            }

            i = right + 1
            
            while (right < s.length - 1 && left > 0 && s(right+1) == s(left-1)) {
                right += 1
                left -= 1
            }

            val newLen = right + 1 - left
            if (newLen > maxLen) {
                maxLen = newLen
                start = left
            }
        }

        return s.slice(start, start+maxLen)
    }

    def main(args: Array[String]): Unit = {
        val s = "babad"
        println(longestPalindrome(s))
    }
}